### PR TITLE
Add parallel downloads job

### PR DESCRIPTION
**Pull request by sethiay**
### Description
Add parallel downloads job

Note: I will add more unit/composite tests in the next PR. Also, I will enable the configs to run e2e tests with parallel downloads in a separate PR.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: internal/cache/file/downloader/parallel_downloads_job.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+		select {
+		case <-job.cancelCtx.Done():
+			return
+		default:
+			if start < end {
+				// Parallely download different ranges not more than maxSingleStepReadSize
+				// and not using go routines more than job.downloadParallelism
+				downloadErrGroup, downloadErrGroupCtx := errgroup.WithContext(job.cancelCtx)
+
+				var singleStepReadSize int64 = 0
```
Comment (raj-prince) [inline internal/cache/file/downloader/parallel_downloads_job.go:120]: Why singleStepReadSize and maxSingleStepReadSize variables are required? Anyway we are checking the end case using (start < end), right?

File: internal/cache/file/downloader/parallel_downloads_job_test.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
```
Comment (raj-prince) [inline internal/cache/file/downloader/parallel_downloads_job_test.go:1]: nit: 2024

File: internal/cache/file/downloader/job.go [hunk 3]
```diff
@@ -269,26 +268,33 @@ func (job *Job) updateFileInfoCache() (err error) {
 	return
 }
 
+// cleanUpDownloadAsyncJob is a helper function which performs clean up tasks
+// for the async job and this should be called at the end of async job.
+//
+// Acquires and releases LOCK(job.mu)
+func (job *Job) cleanUpDownloadAsyncJob() {
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:275]: Should we write unit test for this?

File: internal/cache/file/downloader/parallel_downloads_job.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/storage/gcs"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/util"
+	"golang.org/x/sync/errgroup"
+)
+
+// downloadRangeToFile is a helper function to download a given range of object
+// from GCS into given file handle.
+//
+// This function doesn't take locks and can be executed parallely.
+func (job *Job) downloadRangeToFile(ctx context.Context, cacheFile *os.File, start, end int64) error {
```
Comment (raj-prince) [inline internal/cache/file/downloader/parallel_downloads_job.go:38]: Should we pass the offsetWriter here?

**Review by raj-prince — COMMENTED** [review]
A couple of minor comments.

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 4 above)
Comment (raj-prince) [inline internal/cache/file/downloader/parallel_downloads_job.go:38]: Also, assuming unit test for this method will be in the next separate PR.

Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go:38]: Trying to offload some work from `parallelDownloadObjectAsync` as that's already long and allow unit testing that separately.

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 1 above)
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go:120]: So, this check is to ensure that in one step (includes writing to file info cache), we only write chunk size * download-parallelism-per-file content only whereas start and end are start and end of file

File: internal/cache/file/downloader/job.go (hunk 3 above)
Comment (sethiay) [reply internal/cache/file/downloader/job.go:275]: done

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 1 above)
Comment (vadlakondaswetha) [reply internal/cache/file/downloader/parallel_downloads_job.go:120]: Isn;t this simple loop sufficient:


for (ind = 0; ind < dowloadParallelism && start < end; ind ++) {
   range_start = start 
   range_end = min(end, range_start+chunk_size) // chunk_size tell how much each go routine should download
   start = range_end
}

File: internal/cache/file/downloader/parallel_downloads_job.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+						return job.downloadRangeToFile(downloadErrGroupCtx, cacheFile, rangeStart, rangeEnd)
+					})
+
+					singleStepReadSize = singleStepReadSize + rangeEnd - rangeStart
+					start = rangeEnd
+				}
+				// If any of the go routine failed, consider the async job failed.
+				err = downloadErrGroup.Wait()
+				if err != nil {
+					if errors.Is(err, context.Canceled) {
```
Comment (vadlakondaswetha) [inline internal/cache/file/downloader/parallel_downloads_job.go:136]: why is context canceled means invalid job?

File: internal/cache/file/downloader/parallel_downloads_job.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+					return
+				}
+				job.mu.Unlock()
+				// Change status of job in case of error while updating file cache.
+				if err != nil {
+					job.failWhileDownloading(err)
+					return
+				}
+			} else {
+				err = job.validateCRC()
```
Comment (vadlakondaswetha) [inline internal/cache/file/downloader/parallel_downloads_job.go:168]: why are we duplicating the code so much?

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 1 above)
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go:120]: As discussed, I will make this change

File: internal/cache/file/downloader/parallel_downloads_job.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+func (job *Job) parallelDownloadObjectAsync() {
+	// Cleanup the async job in all cases - completion/failure/invalidation.
+	defer job.cleanUpDownloadAsyncJob()
+
+	// Create, and open cache file for writing object into it.
+	cacheFile, err := cacheutil.CreateFile(job.fileSpec, os.O_TRUNC|os.O_RDWR)
+	if err != nil {
+		err = fmt.Errorf("parallelDownloadObjectAsync: error in creating cache file: %w", err)
+		job.failWhileDownloading(err)
+		return
```
[the remark this answers is not in the document]
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go:88]: I think both the logic are already small to be created a separate function for. If you think creating separate ones improve readability then I will.

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 5 above)
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go:136]: the context passed is the context associated with the job and in the current job flows, context is canceled only when the job is Invalidated (during eviction).

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 6 above)
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go:168]: I tried to but it's not straightforward to break down the function into logically independent blocks. 
Trying one more thing, hopefully that should (but would require additional unit testing)

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vadlakondaswetha) [reply internal/cache/file/downloader/parallel_downloads_job.go:88]: we need to create seperate functions.

File: internal/cache/file/downloader/parallel_downloads_job.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	var start, end int64
+	end = int64(job.object.Size)
+	var parallelReadRequestSize = int64(job.fileCacheConfig.ReadRequestSizeMB) * cacheutil.MiB
+	var maxSingleStepReadSize = parallelReadRequestSize * int64(job.fileCacheConfig.DownloadParallelismPerFile)
+
+	for {
+		select {
+		case <-job.cancelCtx.Done():
+			return
+		default:
```
Comment (vadlakondaswetha) [inline internal/cache/file/downloader/parallel_downloads_job.go:114]: Looking at the code more i see only the default clause is different for sequential download vs parallel download, why can;t we put a if -else based on flag and invoke seperate functions?

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 4 above)
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go]: Done. PTAL

File: internal/cache/file/downloader/parallel_downloads_job.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+		err = fmt.Errorf("downloadRangeToFile: error at the time of copying content to cache file %w", err)
+	}
+	return err
+}
+
+// parallelDownloadObjectAsync does parallel download of the backing GCS object
+// into a file as part of file cache using multiple NewReader method of
+// gcs.Bucket running in parallel
+//
+// Note: There can only be one async parallel or non-parallel download running
```
[the remark this answers is not in the document]
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go]: Removed.

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go]: refactored

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 6 above)
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go]: refactored. Will add unit tests in next PR for refactored methods.

File: internal/cache/file/downloader/parallel_downloads_job.go (hunk 8 above)
Comment (sethiay) [reply internal/cache/file/downloader/parallel_downloads_job.go]: done. PTAL

File: internal/cache/file/downloader/job.go [hunk 10]
```diff
@@ -226,196 +225,186 @@ func (job *Job) notifySubscribers() {
...
 	}
-
-	logger.Tracef("Job:%p (%s:/%s) downloaded till %v offset.", job, job.bucket.Name(), job.object.Name, job.status.Offset)
+	job.mu.Lock()
+	defer job.mu.Unlock()
 	err = job.fileInfoCache.UpdateWithoutChangingOrder(fileInfoKeyName, updatedFileInfo)
-	if err != nil {
-		err = fmt.Errorf("updateFileInfoCache: error while inserting into fileInfoCache %s: %w", updatedFileInfo.Key, err)
-		return
+	if err == nil {
```
Comment (vadlakondaswetha) [inline internal/cache/file/downloader/job.go:267]: lets use the same format everywhere
if err != nil {
  // do something
return
}

continue with other logic

File: internal/cache/file/downloader/job.go [hunk 11]
```diff
@@ -226,196 +225,186 @@ func (job *Job) notifySubscribers() {
...
+		start += maxRead
+		if start == newReaderLimit {
+			err = newReader.Close()
+			if err != nil {
+				logger.Errorf("Job:%p (%s:/%s) error while closing reader: %v", job, job.bucket.Name(), job.object.Name, err)
+			}
+			newReader = nil
+		}
+
+		err = job.updateStatusOffset(start)
```
Comment (vadlakondaswetha) [inline internal/cache/file/downloader/job.go:331]: Please add a comment why are we not returning from the above err.

File: internal/cache/file/downloader/job.go [hunk 12]
```diff
@@ -226,196 +225,186 @@ func (job *Job) notifySubscribers() {
...
+					Range: &gcs.ByteRange{
+						Start: uint64(start),
+						Limit: uint64(newReaderLimit),
+					},
+					ReadCompressed: job.object.HasContentEncodingGzip(),
+				})
+			if err != nil {
+				err = fmt.Errorf("downloadObjectToFile: error in creating NewReader with start %d and limit %d: %w", start, newReaderLimit, err)
+				return err
+			}
```
Comment (vadlakondaswetha) [inline internal/cache/file/downloader/job.go:308]: can you extract this into seperate function so it can be reused in parallel_downloads_job also

File: internal/cache/file/downloader/job.go [hunk 13]
```diff
@@ -226,196 +225,186 @@ func (job *Job) notifySubscribers() {
...
 	}
 
-	var newReader io.ReadCloser
-	var start, end, sequentialReadSize, newReaderLimit int64
-	end = int64(job.object.Size)
-	sequentialReadSize = int64(job.sequentialReadSizeMb) * cacheutil.MiB
-
-	for {
-		select {
-		case <-job.cancelCtx.Done():
```
Comment (vadlakondaswetha) [inline internal/cache/file/downloader/job.go:322]: where are we checking if context is cancelled?

Comment (sethiay) [reply internal/cache/file/downloader/job.go:322]: It's not required because both parallelDownloadObjectToFile & downloadObjectToFile returns err (context.Canceled type) if context is canceled. This is because the `newReader` that we get as output of NewReader call is derived from job.cancelCtx and hence `_, err = io.CopyN(offsetWriter, newReader, maxRead)` fails in that case with context canceled error. This is equivalent to what was there before.

File: internal/cache/file/downloader/job.go (hunk 12 above)
Comment (sethiay) [reply internal/cache/file/downloader/job.go:308]: Could you elaborate what code to put in separate function ?

Assuming you are asking to put {NewReader + copy} logic into a function and then use that in both parallel and non parallel downloads. If yes, then I would avoid that because:
1. NewReader and Copy functions are already abstracted enough and are only called at both the places with their returned errors handled.
2. The read + copy logic is not common in parallel and non parallel downloads. In former case whatever is read is copied to file whereas in latter, a read request of 200MiB is made and data is copied in chunks of 8MiB with updateOffset being called.
3. Even if we go ahead with common function, it's signature will be something like downloadRange(ctx context.Context, dstWriter io.Writer, start int64, end int64, readRequestSize int64, copyChunkSize int64, debugReadType string). This will remove slight duplicacy at the cost of many arguments. 
So, I would prefer keeping it as it is but lmk.

In case you are asking to just abstract out NewReader + err handling code, then also I would prefer keeping it as it is.

File: internal/cache/file/downloader/job.go (hunk 11 above)
Comment (sethiay) [reply internal/cache/file/downloader/job.go:335]: Added. PTAL

File: internal/cache/file/downloader/job.go [hunk 14]
```diff
@@ -275,130 +365,52 @@ func (job *Job) updateFileInfoCache() (err error) {
...
-			job.failWhileDownloading(err)
+			job.handleError(err)
 		}
 	}()
 
-	var newReader io.ReadCloser
-	var start, end, sequentialReadSize, newReaderLimit int64
-	end = int64(job.object.Size)
-	sequentialReadSize = int64(job.sequentialReadSizeMb) * cacheutil.MiB
+	// Both parallel and non-parallel download functions supports cancellation in
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:386]: nit: download functions `support`

File: internal/cache/file/downloader/job_test.go [hunk 15]
```diff
@@ -89,6 +89,11 @@ func (dt *downloaderTest) initJobTest(objectName string, objectContent []byte, s
 	AssertEq(nil, err)
 }
 
+func (dt *downloaderTest) verifyInvalidError(err error) {
```
Comment (raj-prince) [inline internal/cache/file/downloader/job_test.go:92]: Optional: but I would prefer to duplicate for more readability.

File: internal/cache/file/downloader/job.go [hunk 16]
```diff
@@ -226,47 +225,138 @@ func (job *Job) notifySubscribers() {
...
-	return
+
+	err = fmt.Errorf("updateStatusOffset: error while updating offset: %v in fileInfoCache %s: %w", downloadedOffset, updatedFileInfo.Key, err)
+	return err
+}
+
+// downloadObjectToFile downloads the backing object from GCS into the given
+// file and updates the file info cache. It uses gcs.Bucket's NewReader method
+// to download the object.
+func (job *Job) downloadObjectToFile(cacheFile *os.File) (err error) {
```
Comment (raj-prince) [inline internal/cache/file/downloader/job.go:282]: Are we writing unit test for this?

**Review by raj-prince — COMMENTED** [review]
Few nits.

LGTM, assuming we are adding unit tests for newly added method as separate PR.

File: internal/cache/file/downloader/job.go (hunk 16 above)
Comment (sethiay) [reply internal/cache/file/downloader/job.go:282]: Yes, will be there in next PR. Not a blocker for this PR since we already have unit tests for Download function and didn't want to make the PR longer.

**Review by vadlakondaswetha — APPROVED** [review]
